# Assignment 4
Barrett Poth, Kevin Wurangian, Bingqing Li, Fan Hong, Yinan Yang

In [1]:
import nltk
from nltk.corpus import stopwords
#nltk.download('stopwords')
from collections import Counter
from sklearn.feature_extraction.text import TfidfVectorizer,CountVectorizer,TfidfTransformer
from sklearn.linear_model import LogisticRegression
import re

In [4]:
import numpy as np
import pandas as pd
pd.set_option('display.max_colwidth', -1)

In [3]:
stopwords = stopwords.words('english')[:]

In [4]:
pd.set_option("display.max_rows",270)

# Task A

### Engagement Score

In [7]:
post = pd.read_csv("insta_data_with_labels.csv").iloc[:,1:]
post.drop(columns='Unnamed: 0.1', inplace=True)
post.head()

urls  \
0  https://scontent-sea1-1.cdninstagram.com/vp/7831adc9c2d624937ec8197b4affcded/5C76DA0B/t51.2885-15/e15/c181.0.718.718/s480x480/44904753_194498844802328_8238377254064471543_n.jpg?ig_cache_key=MTkxMjcyNTE4Njk1Njg2MzM4Mg%3D%3D.2.c   
1  https://scontent-sea1-1.cdninstagram.com/vp/53f7539662906933bbbce050cd4fdc2c/5BEF2CD7/t51.2885-15/e15/s640x640/45543088_970491189813847_3800824331312248168_n.jpg                                                                    
2  https://scontent-sea1-1.cdninstagram.com/vp/1e4ff18aa09db5069b6e1c6e9e65eac0/5C6F3508/t51.2885-15/e15/s480x480/44577672_336304220507048_1738524933735107944_n.jpg?ig_cache_key=MTkxMjQyNTAwNDk3MzMzNTE0NQ%3D%3D.2                    
3  https://scontent-sea1-1.cdninstagram.com/vp/a43b38e630c9b7b76c93d76517e4882e/5C8A9361/t51.2885-15/e15/s480x480/42697363_1856252364473290_1258265168604898264_n.jpg?ig_cache_key=MTkxMjM1MDIwNDMwMDk5NzU5Mg%3D%3D.2                   
4  https://scontent-sea1-1.cdninstagram.com/vp/d64d5d6cf6b350dd6271216386a27160/5C6B3B71/t51.2885-15/e15/s480x480/44358844_1497691817000978_8239499910752733768_n.jpg?ig_cache_key=MTkxMjMwMjQ0NjUxMjg0NDk0NQ%3D%3D.2                   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                        captions  \
0  Photo by @amivitale | Ntipiyon Nonguta and her son Bernard, one, relax with their neighbors as they make beaded belts for theLoisaba Community Trust in Ewaso Village in Laikipia, Kenya. These women are in a community surrounding Loisaba Conservancy (@loisaba_conservancy) and are being supported and given access to international markets. Revenue from ecotourism is reinvested into neighboring communities like #Ewaso . To get involved and learn more about community based conservation, please follow @loisaba_conservancy and @amivitale. @nature_org @natgeoimagecollection #kenya  #magicalkenya  #whyilovekenya  #northernkenya  #africa  #children  #mothersandsons  #mothers  #empowerment  #conservation  #everydayafrica  #photojournalism  #amivitale                                                                                                                                                                                                                                                                                                                                                                                                                                                                   
1  Video by @amivitale | Wild adult elephants drink and cool off at a watering on the Namunyak Wildlife Conservancy. An elephant’s trunk typically holds around half of a gallon of water, but a big bull elephant can hold as much as 2.5 gallons. After spraying themselves, the elephants will often coat themselves with dust which acts as a bug repellent and sunblock. Namunyak Wildlife Conservancy is home to Reteti Elep

In [6]:
max_like = max(post["likes"])
max_like

1539774

In [7]:
max_like = max(post["likes"])
list_norm_like = []
for p in post.itertuples():
    norm_like = p.likes/max_like
    list_norm_like.append(norm_like)
post["norm_like"] = list_norm_like

max_comment = max(post["comments"])
list_norm_comment = []
for p in post.itertuples():
    norm_comment = p.comments/max_comment
    list_norm_comment.append(norm_comment)
    
post["norm_comment"]= list_norm_comment

In [8]:
engagement_score = []
for p in post.itertuples():
    e_score = .4*p.norm_like+.6*p.norm_comment
    engagement_score.append(e_score)

In [9]:
med_e_score = np.median(engagement_score)

In [12]:
top_quartile = np.percentile(engagement_score, 75)

In [13]:
top_quartile

0.12748719610916837

In [14]:
bot_quartile = np.percentile(engagement_score, 25)

In [15]:
bot_quartile

0.05885066955920289

In [10]:
eng_score = []
for i in range(len(engagement_score)):
    if engagement_score[i]>=med_e_score:
        e_score = 1
    else:
        e_score = 0
    eng_score.append(e_score)
eng_score
post["engagement_score"] = eng_score

In [11]:
pretty_post = post[["comments","likes","engagement_score"]]
pretty_post

comments    likes  engagement_score
0    278       61459    0               
1    362       83092    0               
2    431       136026   0               
3    1934      304484   1               
4    907       413142   1               
5    1495      431621   1               
6    433       253932   0               
7    2903      222257   0               
8    590       100128   0               
9    559       157608   0               
10   1210      301508   0               
11   364       223897   0               
12   2037      443850   1               
13   2063      565934   1               
14   762       215530   0               
15   642       226317   0               
16   661       291040   0               
17   1334      387169   1               
18   1560      221035   0               
19   469       191682   0               
20   613       334072   1               
21   612       325298   1               
22   599       291583   0               
23   797       402345   1               
24   4738      1174215  1               
25   667       298833   0               
26   611       161985   0               
27   1117      447725   1               
28   418       214366   0               
29   2184      289799   0               
30   379       114956   0               
31   598       117591   0               
32   403       127285   0               
33   623       215471   0               
34   2592      948274   1               
35   3009      867752   1               
36   1118      518941   1               
37   989       548724   1               
38   1604      329418   1               
39   361       159099   0               
40   3077      585391   1               
41   373       215233   0               
42   249       92419    0               
43   495       143072   0               
44   737       185936   0               
45   1247      347875   1               
46   995       175864   0               
47   900       434961   1               
48   1145      537087   1               
49   683       160774   0               
50   751       267703   0               
51   619       187708   0               
52   4629      548727   1               
53   2464      314657   1               
54   1515      454881   1               
55   416       126192   0               
56   443       122401   0               
57   922       180828   0               
58   961       214579   0               
59   845       233217   0               
60   2637      143501   0               
61   461       116991   0               
62   991       418235   1               
63   104581    1539774  1               
64   892       309899   0               
65   581       149454   0               
66   1947      518938   1               
67   1386      439217   1               
68   488       147996   0               
69   7844      461167   1               
70   5744      257181   1               
71   8024      358647   1               
72   6557      1195598  1               
73   15315     755065   1               
74   1468      611322   1               
75   1986      518672   1               
76   11579     450738   1               
77   1756      232507   0               
78   1048      471717   1               
79   1061      424204   1               
80   644       143938   0               
81   940       469440   1               
82   430       228329   0               
83   955       302832   0               
84   6868      538705   1               
85   1308      481396   1               
86   1007      160925   0               
87   1013      271691   0               
88   1777      703196   1               
89   428       285835   0               
90   4736      162316   0               
91   474       293610   0               
92   965       478427   1               
93   425       194932   0               
94   1121      468022   1               
95   1926      429463   1               
96   850       272924   0   

# Task B

## Logistic regression

### Labels

In [12]:
post.dropna(axis=0, how='any', inplace=True)

In [13]:
label=post["labels"].values.tolist()
label=" ".join(label)
label = label.lower()        

In [14]:
# tokenize the text 
tokenizer = nltk.tokenize.RegexpTokenizer(r'\w+')
labels = tokenizer.tokenize(label)

In [15]:
filter_labels = Counter(x for x in labels if x not in stopwords and x.isdigit() == False and len(x) != 1 and x.isalpha() == True)

In [16]:
filter_labels.most_common()[:25]

[('sky', 78),
 ('wildlife', 48),
 ('tree', 46),
 ('mammal', 45),
 ('water', 44),
 ('fauna', 39),
 ('mountain', 30),
 ('photography', 30),
 ('grass', 30),
 ('ecosystem', 29),
 ('wilderness', 28),
 ('animal', 28),
 ('organism', 27),
 ('rock', 24),
 ('terrestrial', 24),
 ('recreation', 22),
 ('nature', 22),
 ('girl', 22),
 ('plant', 21),
 ('elephant', 21),
 ('landscape', 20),
 ('bird', 19),
 ('darkness', 18),
 ('snout', 18),
 ('sea', 17)]

In [17]:
natural_objects=["wilderness","organism","autumn","spring","reflections","phenomenon","wood","nature","natrual",'penguins','cattle','species','insect','flower', 'whiskers','amphibian','fauna','elephant','animal','whale','bird','reptile','fish','grass','tree','plant','bird','cats','dog','horse','lions','bear','mammoths','elephants','wildlife','mammal',"tiger","snake"]
human = ["immigrants","population","dancer","male","female",'women','refugee',"snout",'chest','eye','hair','face','child','human',"family","children",'girl',"girls","people",'photojournalism',"community","communityies",'migrants',"man"]
geological  = ["desert","national park",'formation', 'geological','terrain','badlands',"landscape",'mountains',"mountain"'wilderness','forest','highland','grassland','hill','steppe','jungle',"vallay",'horizon','ecosystem','rock','terrestrial','landforms','landform','ecoregion',"geology","hillside"]
aerial =['cloud','sky','sunrise', 'star','atmosphere','aerial','light','darkness', 'sunset','night','astronomical' ,"sun","sunlight"]
activity = ["swimming","marriage","workshop",'festival','cycling','music','fun','climbing','religion','carnival','flame','ceremony','wedding', 'ritual','travel',"paragliding",'game', "recreation",'event','tradition','safari','vacation','sports','sport','competition event', 'boating',"boat"]
artificial_objects = ["pyramids","conservation","artificial",'painting','photography','sanctuary','aircraft',"food",'house','metropolis','outerwear', 'wall','design','room','arts','village','hall','fashion accessory','history', 'furniture','infrastructure', 'campfire','window','building','temple','city','street','vehicle','architecture','leisure','tourism','town','urban','computer','structure',"castle","cities"'town','landmark','town square', 'plaza', 'tourist attraction', 'road',"robots","transportation" ]
water=['island',"islands",'ocean','sea','iceberg', 'arctic ocean', 'shore','sea',"wave",'ice', 'shore','arctic','glacial ', 'polar ice cap', 'ice', 'melting', 'sea', 'ice cap','river','underwater',"coast","snow"]

In [18]:
def replacement(text):
    
    text = text.lower()
    
    for w in human:
        try:
            text = text.replace(w," human ")
        except:
            pass   
    for w in natural_objects:
        try:
            text = text.replace(w, ' natural ')
        except:
            pass
    for w in geological:
        try:
            text = text.replace(w, ' geological ')
        except:
            pass
    for w in aerial:
        try:
            text = text.replace(w, ' aerial ')
        except:
            pass
    for w in activity:
        try:
            text = text.replace(w, ' activity ')
        except:
            pass
    for w in water:
        try:
            text = text.replace(w, ' water ')
        except:
            pass

    for w in artificial_objects:
        try:
            text = text.replace(w, ' artificial ')
        except:
            pass
  

    return text

In [19]:
label_column = []
for p in post['labels']:
    new_text =replacement(p)
    label_column.append(new_text)

In [20]:
def cal_dfitf_score(text,min_df):
    vectorizer=CountVectorizer(stop_words="english", analyzer='word', min_df=min_df, max_features=None)
    transformer=TfidfTransformer()
    tfidf=transformer.fit_transform(vectorizer.fit_transform(text))
    word=vectorizer.get_feature_names()
    weight=tfidf.toarray()
    a=[]
    for i in range(len(weight)):
        a.append(weight[i])
    return a,word

In [21]:
a,word=cal_dfitf_score(label_column,0.1)

In [22]:
df=pd.DataFrame(a)
df.columns=word

In [23]:
df

activity    aerial  artificial  geological        hu     human   natural  \
0    0.544354  0.000000  0.420734    0.000000    0.507174  0.493907  0.159662   
1    0.805181  0.000000  0.000000    0.593030    0.000000  0.000000  0.000000   
2    0.000000  0.228391  0.923573    0.000000    0.000000  0.000000  0.000000   
3    0.000000  0.279709  0.282773    0.808381    0.000000  0.000000  0.214616   
4    0.000000  0.000000  0.000000    0.000000    0.000000  0.000000  0.943419   
5    0.000000  0.246649  0.000000    0.950447    0.000000  0.000000  0.189250   
6    0.385742  0.000000  0.894427    0.000000    0.000000  0.000000  0.226281   
7    0.265003  0.202603  0.819288    0.000000    0.000000  0.000000  0.466361   
8    0.000000  0.000000  0.000000    0.176546    0.000000  0.000000  0.984292   
9    0.000000  0.787009  0.000000    0.505448    0.000000  0.000000  0.000000   
10   0.000000  0.000000  0.118032    0.000000    0.711407  0.692798  0.000000   
11   0.000000  0.000000  0.000000    0.202681    0.000000  0.000000  0.484287   
12   0.000000  0.212267  0.000000    0.000000    0.000000  0.000000  0.977212   
13   0.000000  0.216937  0.000000    0.417977    0.000000  0.000000  0.832261   
14   0.000000  0.000000  0.000000    0.000000    0.000000  0.000000  0.604414   
15   0.000000  0.000000  0.000000    0.000000    0.000000  0.000000  1.000000   
16   0.968378  0.000000  0.249488    0.000000    0.000000  0.000000  0.000000   
17   0.000000  0.404845  0.000000    0.390011    0.000000  0.000000  0.621262   
18   0.000000  0.167997  0.169837    0.971047    0.000000  0.000000  0.000000   
19   0.000000  0.453636  0.000000    0.874029    0.000000  0.000000  0.174034   
20   0.000000  0.145766  0.000000    0.982977    0.000000  0.000000  0.111844   
21   0.000000  0.000000  0.000000    0.000000    0.000000  0.000000  1.000000   
22   0.000000  0.969262  0.195976    0.000000    0.000000  0.000000  0.148740   
23   0.000000  0.000000  0.934944    0.000000    0.000000  0.000000  0.354797   
24   0.000000  0.528668  0.000000    0.848829    0.000000  0.000000  0.000000   
25   0.919227  0.351388  0.177619    0.000000    0.000000  0.000000  0.000000   
26   0.000000  0.000000  0.510897    0.000000    0.615860  0.599750  0.000000   
27   0.705483  0.269681  0.545272    0.000000    0.000000  0.000000  0.000000   
28   1.000000  0.000000  0.000000    0.000000    0.000000  0.000000  0.000000   
29   0.000000  0.000000  0.000000    0.000000    0.716415  0.697674  0.000000   
30   0.000000  0.819914  0.414448    0.394936    0.000000  0.000000  0.000000   
31   0.000000  0.000000  0.000000    0.168547    0.213212  0.207635  0.939692   
32   0.000000  0.453636  0.000000    0.874029    0.000000  0.000000  0.174034   
33   0.000000  0.000000  1.000000    0.000000    0.000000  0.000000  0.000000   
34   0.000000  0.000000  0.000000    0.929042    0.000000  0.000000  0.369975   
35   0.000000  0.139919  0.990163    0.000000    0.000000  0.000000  0.000000   
36   0.330741  0.000000  0.000000    0.000000    0.616301  0.600179  0.388032   
37   0.000000  0.000000  0.234198    0.000000    0.282313  0.274928  0.888744   
38   0.000000  0.427550  0.000000    0.205942    0.000000  0.000000  0.164026   
39   0.000000  0.000000  0.765212    0.000000    0.461212  0.449148  0.000000   
40   0.000000  0.172462  0.000000    0.830716    0.000000  0.000000  0.529309   
41   0.000000  0.000000  0.000000    0.000000    0.716415  0.697674  0.000000   
42   0.000000  0.000000  0.000000    0.155045    0.000000  0.000000  0.987907   
43   0.000000  0.000000  0.000000    0.652961    0.000000  0.000000  0.693415   
44   0.916031  0.000000  0.000000    0.000000    0.213366  0.207785  0.268677   
45   0.000000  0.506788  0.000000    0.000000    0.617600  0.601445  0.000000   
46   0.000000  0.000000  0.000000    0.000000    0.653053  0.635971  0.411172   
47   0.000000  0.000000  0.000000    0.184647    0.467158  0.454938  0.735324   
48   0.000000  0.000000  0.000000   

In [24]:
df.drop(['hu'],axis=1,inplace=True)

In [25]:
y=post["engagement_score"]

In [26]:
from sklearn.model_selection import train_test_split
from sklearn.metrics import confusion_matrix, accuracy_score
def logictic_pred(df,y):
    X_train, X_test, y_train, y_test = train_test_split(df, y, test_size=0.2, random_state=0)
    lr = LogisticRegression() 
    lr.fit(X_train, y_train)
    y_pred=lr.predict(X_test)
    print("confusion_matrix:\n",confusion_matrix(y_test, y_pred))
    print("accuracy_score:\n",accuracy_score(y_test, y_pred))

In [27]:
logictic_pred(df,y)

confusion_matrix:
 [[14  9]
 [ 2 23]]
accuracy_score:
 0.7708333333333334


### Captions

In [28]:
text=post["captions"].values.tolist()
text=" ".join(text)
text = text.lower()
tokens = tokenizer.tokenize(text)

In [29]:
filter_words = Counter(x for x in tokens if x not in stopwords and x.isdigit() == False and len(x) != 1 and x.isalpha() == True)

In [30]:
filter_words.most_common()[:300]

[('photo', 260),
 ('one', 89),
 ('see', 80),
 ('people', 80),
 ('follow', 77),
 ('years', 74),
 ('world', 73),
 ('natgeo', 65),
 ('many', 57),
 ('year', 56),
 ('like', 54),
 ('first', 53),
 ('wildlife', 50),
 ('time', 49),
 ('work', 47),
 ('life', 46),
 ('national', 45),
 ('new', 44),
 ('species', 41),
 ('part', 41),
 ('old', 40),
 ('around', 40),
 ('park', 40),
 ('back', 39),
 ('story', 38),
 ('china', 36),
 ('way', 35),
 ('still', 34),
 ('long', 34),
 ('wild', 34),
 ('women', 33),
 ('get', 32),
 ('conservation', 32),
 ('land', 32),
 ('made', 32),
 ('animals', 31),
 ('last', 29),
 ('day', 29),
 ('make', 28),
 ('also', 28),
 ('found', 28),
 ('kenya', 27),
 ('africa', 27),
 ('today', 27),
 ('images', 27),
 ('nature', 27),
 ('across', 27),
 ('find', 27),
 ('city', 25),
 ('war', 25),
 ('poaching', 25),
 ('live', 25),
 ('project', 25),
 ('assignment', 25),
 ('since', 25),
 ('children', 24),
 ('home', 24),
 ('elephants', 24),
 ('red', 24),
 ('water', 23),
 ('visit', 23),
 ('along', 22),
 ('

In [31]:
def clean_data(text):
    word=[]
    #delete punctuations stopwords
    words = re.compile(r'[^a-zA-Z]+').split(text)
    for w in words:
        if w not in stopwords and len(w)>2:
            word.append(w)
    return word

In [32]:
caption_column = []
for p in post['captions']:
    a=clean_data(p)
    a= ",".join(a)
    new_text =replacement(a) 
    caption_column.append(new_text)

In [33]:
b,w=cal_dfitf_score(caption_column,0.2)

In [34]:
df1=pd.DataFrame(b)
df1.columns=w

In [35]:
df1.drop(["hu","follow","years"],axis=1,inplace=True)

In [36]:
logictic_pred(df1,y)

confusion_matrix:
 [[17  6]
 [ 7 18]]
accuracy_score:
 0.7291666666666666


### labels+captions

In [37]:
post["all"]=post['captions']+post['labels']

In [38]:
all_column = []
for p in post['all']:
    a=clean_data(p)
    a= ",".join(a)
    new_text =replacement(a) 
    all_column.append(new_text)

In [39]:
c,z=cal_dfitf_score(all_column,0.2)

In [40]:
df2=pd.DataFrame(c)
df2.columns=z

In [41]:
df2.drop(["hu","follow","years","photo","world","like","natgeo"],axis=1,inplace=True)

In [42]:
logictic_pred(df2,y)

confusion_matrix:
 [[15  8]
 [ 1 24]]
accuracy_score:
 0.8125


# Task C

In [16]:
post["engagement_score"] = engagement_score

In [17]:
eng_score = []
for i in range(len(engagement_score)):
    if engagement_score[i]>=med_e_score:
        e_score = 1
    else:
        e_score = 0
    eng_score.append(e_score)
eng_score
post["engagement"] = eng_score

In [145]:
# use relevant columns - labels and engagement_score
post_clean = post.drop(post.columns[[0,1,2,3,4,6,7,9]], axis = 1)

In [21]:
post_clean.head()

labels  \
0  ['people', 'public space', 'community', 'tribe', 'tree', 'travel', 'temple', 'recreation', 'tourism', 'crowd']                                                   
1  NaN                                                                                                                                                              
2  ['air sports', 'paragliding', 'landscape', 'sand', 'ecoregion', 'desert', 'powered paragliding', 'windsports', 'sahara', 'steppe']                               
3  ['waterway', 'city', 'cityscape', 'reflection', 'sky', 'landmark', 'river', 'town', 'evening', 'dusk']                                                           
4  ['mountain range', 'mountainous landforms', 'massif', 'nunatak', 'geological phenomenon', 'arête', 'piste', 'aerial photography', 'glacial landform', 'ridge']   

   engagement_score  
0  0.017561          
1  0.023662          
2  0.037809          
3  0.090194          
4  0.112529

In [146]:
post_clean = post_clean.dropna()

In [147]:
post_clean.shape

(238, 2)

In [148]:
post_clean

labels  \
0    ['people', 'public space', 'community', 'tribe', 'tree', 'travel', 'temple', 'recreation', 'tourism', 'crowd']                                                                                  
2    ['air sports', 'paragliding', 'landscape', 'sand', 'ecoregion', 'desert', 'powered paragliding', 'windsports', 'sahara', 'steppe']                                                              
3    ['waterway', 'city', 'cityscape', 'reflection', 'sky', 'landmark', 'river', 'town', 'evening', 'dusk']                                                                                          
4    ['mountain range', 'mountainous landforms', 'massif', 'nunatak', 'geological phenomenon', 'arête', 'piste', 'aerial photography', 'glacial landform', 'ridge']                                  
5    ['marine mammal', 'whales dolphins and porpoises', 'mammal', 'water', 'marine biology', 'killer whale', 'dolphin', 'short beaked common dolphin', 'stenella', 'organism']                       
6    ['highland', 'mountain', 'mountainous landforms', 'ridge', 'badlands', 'sky', 'wilderness', 'fell', 'escarpment', 'national park']                                                              
7    ['black and white', 'infrastructure', 'crowd', 'monochrome photography', 'street', 'monochrome', 'temple', 'religion']                                                                          
9    ['plant', 'tree', 'recreation', 'city', 'leisure', 'park', 'vehicle', 'sky', 'grass', 'tourism']                                                                                                
10   ['ecosystem', 'tree', 'vegetation', 'wildlife', 'nature reserve', 'fauna', 'wilderness', 'shrubland', 'woodland', 'grass']                                                                      
11   ['sky', 'horizon', 'sea', 'calm', 'morning', 'dawn', 'loch', 'sunrise', 'atmosphere', 'highland']                                                                                               
12   ['beauty', 'girl', 'eye', 'brown hair', 'long hair', 'black hair', 'model', 'portrait', 'portrait photography', 'photo shoot']                                                                  
13   ['ecosystem', 'green', 'underwater', 'marine biology', 'fauna', 'organism', 'seaweed', 'fish', 'water', 'algae']                                                                                
15   ['fauna', 'bird', 'beak', 'wildlife', 'wood', 'organism', 'darkness', 'tree']                                                                                                                   
16   ['reflection', 'water', 'sky', 'cattle like mammal', 'wildlife', 'tree', 'landscape', 'bank', 'rock', 'plant']                                                                                  
17   ['water', 'marine mammal', 'marine biology', 'whales dolphins and porpoises', 'fin', 'organism', 'underwater', 'whale', 'sea']                                                                  
18   ['cat', 'small to medium sized cats', 'cat like mammal', 'whiskers', 'kitten', 'fur', 'carnivoran']                                                                                             
19   ['hammock', 'leisure', 'vacation', 'fun', 'recreation']                                                                                                                                         
20   ['wilderness', 'winter', 'geological phenomenon', 'sky', 'leaf', 'morning', 'tundra', 'snow', 'frost', 'freezing']                                                                              
21   ['rock', 'sky', 'badlands', 'mountain', 'formation', 'landscape', 'geology', 'national park', 'archaeological site', 'stock photography']                                                       
22   ['sky', 'cloud', 'mountainous landforms', 'highland', 'mountain', 'hill station', 'hill', 'tree', 'morning', 'mountain range']                                                                  
23   ['rock', 'sky', 'badlands

In [149]:
# clean up
def clean_up(label):
    label = label.replace("[", "")
    label = label.replace("]", "")
    label = label.replace("'", "")
    label = label.replace(",", " ")
    return label

In [150]:
# convert to series and clean up the labels column
post_clean["labels"] = post_clean.iloc[:,0].apply(clean_up)

In [151]:
post_clean

labels  \
0    people  public space  community  tribe  tree  travel  temple  recreation  tourism  crowd                                                                                   
2    air sports  paragliding  landscape  sand  ecoregion  desert  powered paragliding  windsports  sahara  steppe                                                               
3    waterway  city  cityscape  reflection  sky  landmark  river  town  evening  dusk                                                                                           
4    mountain range  mountainous landforms  massif  nunatak  geological phenomenon  arête  piste  aerial photography  glacial landform  ridge                                   
5    marine mammal  whales dolphins and porpoises  mammal  water  marine biology  killer whale  dolphin  short beaked common dolphin  stenella  organism                        
6    highland  mountain  mountainous landforms  ridge  badlands  sky  wilderness  fell  escarpment  national park                                                               
7    black and white  infrastructure  crowd  monochrome photography  street  monochrome  temple  religion                                                                       
9    plant  tree  recreation  city  leisure  park  vehicle  sky  grass  tourism                                                                                                 
10   ecosystem  tree  vegetation  wildlife  nature reserve  fauna  wilderness  shrubland  woodland  grass                                                                       
11   sky  horizon  sea  calm  morning  dawn  loch  sunrise  atmosphere  highland                                                                                                
12   beauty  girl  eye  brown hair  long hair  black hair  model  portrait  portrait photography  photo shoot                                                                   
13   ecosystem  green  underwater  marine biology  fauna  organism  seaweed  fish  water  algae                                                                                 
15   fauna  bird  beak  wildlife  wood  organism  darkness  tree                                                                                                                
16   reflection  water  sky  cattle like mammal  wildlife  tree  landscape  bank  rock  plant                                                                                   
17   water  marine mammal  marine biology  whales dolphins and porpoises  fin  organism  underwater  whale  sea                                                                 
18   cat  small to medium sized cats  cat like mammal  whiskers  kitten  fur  carnivoran                                                                                        
19   hammock  leisure  vacation  fun  recreation                                                                                                                                
20   wilderness  winter  geological phenomenon  sky  leaf  morning  tundra  snow  frost  freezing                                                                               
21   rock  sky  badlands  mountain  formation  landscape  geology  national park  archaeological site  stock photography                                                        
22   sky  cloud  mountainous landforms  highland  mountain  hill station  hill  tree  morning  mountain range                                                                   
23   rock  sky  badlands  geological phenomenon  landscape  geology  formation  mountain  terrain  butte                                                                        
24   wood  hardwood  wood stain                                                                                                                                                 
25   light  lighting  light fixture  darkness  column  wood  night  window  chapel  place of wor

In [152]:
# throw these string of labels into a list
final_labels = post_clean["labels"].tolist()

In [153]:
final_labels

['people  public space  community  tribe  tree  travel  temple  recreation  tourism  crowd',
 'air sports  paragliding  landscape  sand  ecoregion  desert  powered paragliding  windsports  sahara  steppe',
 'waterway  city  cityscape  reflection  sky  landmark  river  town  evening  dusk',
 'mountain range  mountainous landforms  massif  nunatak  geological phenomenon  arête  piste  aerial photography  glacial landform  ridge',
 'marine mammal  whales dolphins and porpoises  mammal  water  marine biology  killer whale  dolphin  short beaked common dolphin  stenella  organism',
 'highland  mountain  mountainous landforms  ridge  badlands  sky  wilderness  fell  escarpment  national park',
 'black and white  infrastructure  crowd  monochrome photography  street  monochrome  temple  religion',
 'plant  tree  recreation  city  leisure  park  vehicle  sky  grass  tourism',
 'ecosystem  tree  vegetation  wildlife  nature reserve  fauna  wilderness  shrubland  woodland  grass',
 'sky  horizon

In [154]:
# init count vectorizer
vectorizer = CountVectorizer(analyzer='word',       
                             stop_words='english',             # remove stop words
                             lowercase=True,                   # convert all words to lowercase
                            )

In [155]:
data_vectorized = vectorizer.fit_transform(final_labels)

## LDA model with 6 topics

In [156]:
lda = LatentDirichletAllocation(n_components = 6, random_state = 0)

In [157]:
lda.fit(data_vectorized)

c:\python\python37\lib\site-packages\sklearn\decomposition\online_lda.py:536: DeprecationWarning: The default value for 'learning_method' will be changed from 'online' to 'batch' in the release 0.20. This warning was introduced in 0.18.
  DeprecationWarning)


LatentDirichletAllocation(batch_size=128, doc_topic_prior=None,
             evaluate_every=-1, learning_decay=0.7, learning_method=None,
             learning_offset=10.0, max_doc_update_iter=100, max_iter=10,
             mean_change_tol=0.001, n_components=6, n_jobs=1,
             n_topics=None, perp_tol=0.1, random_state=0,
             topic_word_prior=None, total_samples=1000000.0, verbose=0)

In [158]:
lda.n_components

6

## Which Words Load on Which Topics (Weights of Each Keyword in Each Topic)

In [159]:
# column names
topicnames = ["Topic" + str(i) for i in range(1, lda.n_components + 1)]

# weights of each keyword in each topic

# Topic-Keyword Matrix
df_topic_keywords = pd.DataFrame(lda.components_.T)

# Assign Column and Index
df_topic_keywords.index = vectorizer.get_feature_names()
df_topic_keywords.columns = topicnames

# View
df_topic_keywords

Topic1     Topic2     Topic3     Topic4    Topic5  \
abdomen          0.232944  0.215697   0.226181   0.219692   1.281541   
accessory        1.054471  2.111974   1.283405   0.217115   0.221841   
accipitriformes  0.233699  0.221102   0.228149   1.062371   0.223267   
adventure        0.214649  1.692356   0.213931   0.334155   0.213432   
advertising      0.220114  0.228623   0.219820   1.040164   0.224346   
aeolian          0.223491  0.225654   2.037220   0.251282   0.218932   
aerial           0.452700  0.279628   1.266694   3.484535   0.215822   
african          5.876329  0.213844   0.214889   0.226692   0.211241   
agriculture      0.223709  2.103468   0.218017   0.233473   0.215909   
air              0.222980  0.217158   2.115748   0.209445   0.206544   
aircraft         0.231036  0.218811   1.269603   0.216252   0.213253   
algae            0.247249  0.229377   0.221700   1.029066   0.217294   
alps             0.205529  1.279580   0.224074   0.215228   0.231285   
american         0.253990  0.225464   0.211892   2.302694   0.219583   
amphibian        1.047051  0.211506   0.216802   0.243792   0.225627   
ancient          0.211759  0.219681   0.336213   0.238183   2.814510   
animal           8.030207  0.251147   0.259622   18.093572  1.300346   
ape              0.230815  0.229773   0.224566   1.276017   0.218775   
arch             0.215157  0.232005   0.215318   0.208334   2.100424   
archaeological   0.208733  0.231982   0.349077   0.223524   2.309666   
archipelago      1.031316  0.216920   0.226200   0.250316   0.216899   
architecture     0.229005  0.394903   2.182571   0.228530   4.366636   
arctic           0.214216  8.171616   1.274290   0.850247   0.253015   
area             0.211998  0.477259   9.610116   0.373508   0.231656   
arecales         0.217507  1.052017   0.215837   0.209875   0.215102   
arena            0.217964  0.219937   0.218836   1.040133   0.211891   
arm              0.210822  0.219013   0.233178   0.226668   1.283495   
arranging        0.219388  0.222183   0.219440   0.220185   0.232751   
art              0.215133  0.221295   0.230476   0.222539   2.104894   
arts             0.223343  0.211607   0.214699   0.218001   2.112568   
arête            0.220587  0.232798   0.216959   1.034201   0.217475   
asian            0.223592  0.225863   0.224522   0.221222   1.283575   
astronomical     0.216735  0.219042   0.232752   0.217611   0.217975   
atmosphere       0.246365  0.265261   5.275460   1.661817   0.253374   
attraction       0.215639  0.229921   2.189051   0.229199   3.071911   
autumn           0.210188  2.694548   1.088851   0.219712   0.218113   
aviation         0.210748  0.219541   1.277359   0.228865   0.218402   
badlands         0.251750  1.315767   2.548015   0.829389   0.221769   
ball             0.225626  1.279814   0.217033   0.225153   0.212199   
bank             0.221624  0.221134   0.226794   2.935678   0.241749   
barechestedness  0.223494  0.222784   1.286154   0.217117   1.270321   
basketball       0.216416  3.398839   0.226695   0.216062   0.209605   
bay              0.218763  0.211190   1.275069   0.221153   0.224093   
beach            0.221137  1.282549   0.223988   0.218099   0.225384   
beak             2.820272  0.253182   0.224799   7.220264   0.242947   
beaked           0.226810  0.242376   0.228639   1.024762   0.218152   
bear             0.342501  0.222968   0.219315   9.596487   0.225969   
beauty           1.042346  0.222355   0.225644   0.218574   1.055640   
bedrock          0.228786  0.247264   0.244018   0.271952   0.231848   
bedroom          0.219621  0.225924   0.217649   0.215006   0.209741   
behavior         0.218202  0.219938   0.234991   0.310295   2.403766   
bicycle          0.229057  0.233361   3.169737   0.215470   0.219381   
big              0.223141  0.220294   0.211367   8.609363   0.206766   
biology          0.449039  0.230342   0.279604   6.953377   0.217185   
biome            0.221611  0.227449   0.219635   1.039451  

In [160]:
# throw this onto a csv
df_topic_keywords.to_csv("wordstotopics2.csv")

In [177]:
# Topic 1 
topic1 = pd.DataFrame(df_topic_keywords.sort_values(by = ['Topic1'], ascending = False).head().index.values)
topic1.columns = ["Topic1"]
topic1

Topic1
0  elephant   
1  photography
2  animal     
3  mammoths   
4  elephants

We decided to name this topic "Big Mammals"

In [178]:
# topic 2
topic2 = pd.DataFrame(df_topic_keywords.sort_values(by = ['Topic2'], ascending = False).head().index.values)
topic2.columns = ["Topic2"]
topic2

Topic2
0  tree      
1  sky       
2  recreation
3  water     
4  mountain

We decided to name this topic "Nature/Landscape"

In [180]:
# topic 3
topic3 = pd.DataFrame(df_topic_keywords.sort_values(by = ['Topic3'], ascending = False).head().index.values)
topic3.columns = ["Topic3"]
topic3

Topic3
0  sky      
1  ecosystem
2  grassland
3  area     
4  city

We decided to name this topic "ecosystem"

In [181]:
# topic 4
topic4 = pd.DataFrame(df_topic_keywords.sort_values(by = ['Topic4'], ascending = False).head().index.values)
topic4.columns = ["Topic4"]
topic4

Topic4
0  wildlife
1  mammal  
2  fauna   
3  organism
4  animal

We decided to name this topic "Wildlife"

In [183]:
# topic 5
topic5 = pd.DataFrame(df_topic_keywords.sort_values(by = ['Topic5'], ascending = False).head().index.values)
topic5.columns = ["Topic5"]
topic5

Topic5
0  human 
1  girl  
2  temple
3  horse 
4  site

We decided to name this topic "People & Exploration"

### Topic Weights For Each Image

In [184]:
# Create Document - Topic Matrix
lda_output = lda.transform(data_vectorized)

# index names
postnames = ["Post" + str(i) for i in range(1, len(final_labels) + 1)]

# Make the pandas dataframe
df_post_topic = pd.DataFrame(np.round(lda_output, 2), columns=topicnames, index=postnames)

# Get dominant topic for each document
dominant_topic = np.argmax(df_post_topic.values, axis=1)
for i in range(len(dominant_topic)):
    dominant_topic[i] = dominant_topic[i] + 1
    
# df_post_topic['dominant_topic'] = dominant_topic

# Styling
def color_green(val):
    color = 'green' if val > .1 else 'black'
    return 'color: {col}'.format(col=color)

def make_bold(val):
    weight = 700 if val > .1 else 400
    return 'font-weight: {weight}'.format(weight=weight)

# Apply Style
df_post_topics = df_post_topic.style.applymap(color_green).applymap(make_bold)
df_post_topics

In [186]:
df_post_topic = df_post_topic.rename(columns={'Topic1': 'Big Mammals', 'Topic2': 'Nature', 'Topic3': 'Ecosystem', 'Topic4':'Wildlife', 'Topic5': 'People & Exploration', 'Topic6':'Terrain'})

In [202]:
post_clean.tail()

labels  \
265  sport venue  man  fan  stadium  structure  player  jersey  product  outerwear  competition event                         
266  reflection  water  waterway  night  tree  city  evening  street  screenshot  sky                                         
267  fauna  wildlife  nature reserve  primate  jungle  terrestrial animal  organism  forest  rainforest  grass                
268  aviation  aircraft  air travel  sky                                                                                      
269  mountain  mountainous landforms  sky  ridge  geological phenomenon  adventure  atmosphere  cloud  terrain  mountaineer   

     engagement_score  
265  0.062915          
266  0.106362          
267  0.085054          
268  0.042053          
269  0.104143

In [209]:
df_post_topic = df_post_topic.reset_index()

In [208]:
post_clean = post_clean.reset_index()

In [210]:
df_post_topic["engagement"] = post_clean["engagement_score"]

In [211]:
df_post_topic

index  Big Mammals  Nature  Ecosystem  Wildlife  People & Exploration  \
0    Post1    0.01         0.93    0.01       0.01      0.01                   
1    Post2    0.01         0.01    0.94       0.01      0.01                   
2    Post3    0.02         0.02    0.92       0.02      0.02                   
3    Post4    0.01         0.43    0.01       0.53      0.01                   
4    Post5    0.01         0.01    0.01       0.96      0.01                   
5    Post6    0.01         0.94    0.01       0.01      0.01                   
6    Post7    0.92         0.02    0.02       0.02      0.02                   
7    Post8    0.02         0.92    0.02       0.02      0.02                   
8    Post9    0.01         0.01    0.01       0.93      0.01                   
9    Post10   0.02         0.02    0.92       0.02      0.02                   
10   Post11   0.95         0.01    0.01       0.01      0.01                   
11   Post12   0.01         0.01    0.01       0.93      0.01                   
12   Post13   0.02         0.02    0.02       0.91      0.02                   
13   Post14   0.01         0.01    0.27       0.68      0.01                   
14   Post15   0.01         0.01    0.01       0.94      0.01                   
15   Post16   0.01         0.01    0.01       0.94      0.01                   
16   Post17   0.03         0.86    0.03       0.03      0.03                   
17   Post18   0.01         0.01    0.33       0.62      0.01                   
18   Post19   0.01         0.01    0.01       0.01      0.01                   
19   Post20   0.01         0.94    0.01       0.01      0.01                   
20   Post21   0.01         0.01    0.01       0.01      0.01                   
21   Post22   0.03         0.03    0.03       0.83      0.03                   
22   Post23   0.01         0.01    0.01       0.01      0.01                   
23   Post24   0.01         0.17    0.78       0.01      0.01                   
24   Post25   0.01         0.01    0.46       0.01      0.01                   
25   Post26   0.02         0.02    0.02       0.02      0.02                   
26   Post27   0.03         0.83    0.03       0.03      0.03                   
27   Post28   0.02         0.27    0.02       0.66      0.02                   
28   Post29   0.01         0.94    0.01       0.01      0.01                   
29   Post30   0.01         0.01    0.01       0.01      0.40                   
30   Post31   0.02         0.02    0.02       0.02      0.02                   
31   Post32   0.01         0.01    0.01       0.94      0.01                   
32   Post33   0.01         0.01    0.94       0.01      0.01                   
33   Post34   0.01         0.01    0.01       0.01      0.94                   
34   Post35   0.01         0.01    0.01       0.94      0.01                   
35   Post36   0.01         0.01    0.94       0.01      0.01                   
36   Post37   0.02         0.02    0.02       0.02      0.92                   
37   Post38   0.61         0.01    0.01       0.34      0.01                   
38   Post39   0.02         0.02    0.92       0.02      0.02                   
39   Post40   0.02         0.02    0.02       0.02      0.90                   
40   Post41   0.01         0.01    0.94       0.01      0.01                   
41   Post42   0.08         0.08    0.08       0.08      0.58                   
42   Post43   0.01         0.01    0.01       0.94      0.01                   
43   Post44   0.01         0.01    0.01       0.93      0.01                   
44   Post45   0.01         0.93    0.01       0.01      0.01                   
45   Post46   0.79         0.04    0.04       0.04      0.04                   
46   Post47   0.03         0.56    0.03       0.33      0.03                   
47   Post48   0.01         0.01    0.01       0.94      0.01                   
48   Post49   0.02         0.02    0.02       0.92      0.02                   
49   Po

In [212]:
top_quartile

0.12748719610916837

In [213]:
bot_quartile

0.05885066955920289

## Top Quartile 

In [216]:
# Top Quartile of engagement 
top_q = df_post_topic[df_post_topic.apply(lambda x: x['engagement'] > top_quartile, axis=1)]
top_q.head()

index  Big Mammals  Nature  Ecosystem  Wildlife  People & Exploration  \
11  Post12  0.01         0.01    0.01       0.93      0.01                   
21  Post22  0.03         0.03    0.03       0.83      0.03                   
31  Post32  0.01         0.01    0.01       0.94      0.01                   
32  Post33  0.01         0.01    0.94       0.01      0.01                   
33  Post34  0.01         0.01    0.01       0.01      0.94                   

    Terrain  engagement  
11  0.01     0.158853    
21  0.03     0.332218    
31  0.01     0.261212    
32  0.01     0.242686    
33  0.01     0.141224

In [232]:
top_q = []

In [233]:
top_q_list.append(top_q["Big Mammals"].mean())

In [234]:
top_q_list.append(top_q["Nature"].mean())

In [235]:
top_q_list.append(top_q["Ecosystem"].mean())

In [236]:
top_q_list.append(top_q["Wildlife"].mean())

In [237]:
top_q_list.append(top_q["People & Exploration"].mean())

In [238]:
top_q_list.append(top_q["Terrain"].mean())

## Bot Quartile

In [223]:
# Bot Quartile of engagement 
bot_q = df_post_topic[df_post_topic.apply(lambda x: x['engagement'] < bot_quartile, axis=1)]
bot_q.head()

index  Big Mammals  Nature  Ecosystem  Wildlife  People & Exploration  \
0   Post1   0.01         0.93    0.01       0.01      0.01                   
1   Post2   0.01         0.01    0.94       0.01      0.01                   
7   Post8   0.02         0.92    0.02       0.02      0.02                   
16  Post17  0.03         0.86    0.03       0.03      0.03                   
23  Post24  0.01         0.17    0.78       0.01      0.01                   

    Terrain  engagement  
0   0.01     0.017561    
1   0.01     0.037809    
7   0.02     0.044150    
16  0.03     0.052486    
23  0.01     0.045586

In [239]:
bot_q_list = []

In [240]:
bot_q_list.append(bot_q["Big Mammals"].mean())

In [241]:
bot_q_list.append(bot_q["Nature"].mean())

In [242]:
bot_q_list.append(bot_q["Ecosystem"].mean())

In [243]:
bot_q_list.append(bot_q["Wildlife"].mean())

In [244]:
bot_q_list.append(bot_q["People & Exploration"].mean())

In [245]:
bot_q_list.append(bot_q["Terrain"].mean())

In [254]:
table = ["Big Mammals", "Nature", "Ecosystem", "Wildlife", "People & Exploration", "Terrain"]

In [255]:
final_table = pd.DataFrame(table, columns = ["topics"])

In [256]:
final_table

topics
0  Big Mammals         
1  Nature              
2  Ecosystem           
3  Wildlife            
4  People & Exploration
5  Terrain

In [257]:
final_table["Top_Q"] = top_q_list

In [258]:
final_table["Bot_Q"] = bot_q_list

In [259]:
final_table

topics     Top_Q     Bot_Q
0  Big Mammals           0.163077  0.036034
1  Nature                0.087538  0.352759
2  Ecosystem             0.158308  0.149310
3  Wildlife              0.446615  0.106724
4  People & Exploration  0.044615  0.195172
5  Terrain               0.092923  0.156724

# Task D

##### Task C
We get higher accuracy when predicting engagement scores by using labels alone, rather than captions alone. The reason may be because the captions may give more detailed information about the context behind the pictures, while the labels focus on the content in the pictures. The best results came when combining the labels and captions to predict engagement scores. This makes sense because having both the pictures and their content from the labels, along with possible context and outside information from the captions, makes more of a complete picture in which instagram users engage upon. Our highest accuracy achieved was 0.81.

##### Task D

In the table above the top quartile which had the highest engagement scores had the highest average topic weights on "Wildlife", "Big Mammals", and "Ecosystem". While the bottom quartile which had the lowest engagement scores had the highest average topic weights on "Nature", "People & Exploration", and "Terrain". 

In order to increase engagement, we reccommend that NatGeo consider the strategy to increase posting about "Wildlife", "Big Mammals", and the "Ecosystem" and less emphasis on "Nature", "People & Exploration", and "Terrain". 

It seems that people on the NatGeo instagram page are more engaged with all things related to animals and wildlife, and are less interested in pictures of terrain, landscapes, and other people. 

Another strategy NatGeo should consider is to experiment to have pictures that include both animals and people/landscapes together to stimulate engagement on less popular topics to perhaps tap into hidden opportunities for engagement, connecting the people who engage with animal pictures to also engage in less popular topics. 


